### PR TITLE
BAH-1684 | Functionality to upload report against a panel

### PR DESCRIPTION
**Pull request by Arjun-Go**

### CODE REVIEW

**Review by pajmera4 — APPROVED** [review]

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+import org.openmrs.*;
+import org.openmrs.api.ConceptService;
+import org.openmrs.module.fhirExtension.domain.observation.LabResult;
+import org.openmrs.module.fhirExtension.translators.impl.DiagnosticReportObsTranslatorHelper;
+
+import java.util.*;
```
Comment (N0-man) [inline api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:13]: avoid using wild card imports
https://rules.sonarsource.com/java/RSPEC-2208

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+	private static final int PATIENT_ID = 12345;
+	
+	@Mock
+	private ConceptService conceptService;
+	
+	@InjectMocks
+	private final DiagnosticReportObsTranslatorHelper diagnosticReportObsResultTranslatorHelper = new DiagnosticReportObsTranslatorHelper();
+	
+	@Test
+    public void shouldTranslateToObsWhenTestReportIsUploadedWithLabNotesForSingleTestWithoutAnyOrder() {
```
Comment (N0-man) [inline api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:39]: this name is hard... you can use a separator to help
`givenSingleTest_WhenTestReportIsUploadedWithNotes_ShouldTranslateToObs`

Comment (N0-man) [reply api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:39]: Also does the requirements differ for single vs multiple test? ordered or non ordered test?

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+                case "LAB_REPORT":
+                    assertEquals(REPORT_URL, obsModelChildLevel.getValueText());
+                    break;
+                case "LAB_NOTES":
+                    assertEquals(LAB_TEST_NOTES, obsModelChildLevel.getValueText());
+                    break;
+                case "LAB_RESULT":
+                    assertEquals(REPORT_NAME, obsModelChildLevel.getValueText());
+                    break;
+            }
```
Comment (N0-man) [inline api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:79]: this test would pass even if all the 3 observations are duplicate

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+		assertEquals(1, obsGroupMembersSecondLevel.size());
+		
+		Obs obsModelThirdLevel = obsGroupMembersSecondLevel.iterator().next();
+		assertPatientAndObservationDate(testDate, obsModelSecondLevel);
+		assertEquals("LAB_NOTES", obsModelThirdLevel.getConcept().getPreferredName(ENGLISH).getName());
+		assertEquals(LAB_TEST_NOTES, obsModelThirdLevel.getValueText());
+	}
+	
+	@Test
+    public void shouldTranslateToObsWhenTestReportIsUploadedWithLabNotesForPanelWithoutAnyOrder() {
```
Comment (N0-man) [inline api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:116]: does it matter if the panel comes from order or not?

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+        assertPatientAndObservationDate(testDate, obsModel);
+        assertEquals(panel, obsModel.getConcept());
+        Set<Obs> obsGroupMembersTopLevel = obsModel.getGroupMembers();
+        assertEquals(2, obsGroupMembersTopLevel.size());
+
+        List<Obs> obsGroupMembersTopLevelList = new ArrayList<>();
+
+
+        for (int topLevelIndex = 0; topLevelIndex < obsGroupMembersTopLevelList.size(); topLevelIndex++) {
```
Comment (N0-man) [inline api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:147]: difficult to read...

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java (hunk 1 above)
Comment (Arjun-Go) [reply api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java]: Changes made

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java (hunk 2 above)
Comment (Arjun-Go) [reply api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java]: > this name is hard... you can use a separator to help `givenSingleTest_WhenTestReportIsUploadedWithNotes_ShouldTranslateToObs`

Test name updated.

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java (hunk 3 above)
Comment (Arjun-Go) [reply api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:81]: Yes. Also, no sort of validation present in the frontend. Do we need that as well?

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java (hunk 4 above)
Comment (Arjun-Go) [reply api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:118]: At the moment, no. Nothing unique in the payload if it comes from order.

Comment (N0-man) [reply api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:118]: Then you should consider removing it from your test name

Comment (Arjun-Go) [reply api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:118]: Sure. Will make the change accordingly.

File: api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java (hunk 3 above)
Comment (Arjun-Go) [reply api/src/test/java/org/openmrs/module/fhirExtension/translators/DiagnosticReportObsTranslatorHelperTest.java:81]: This has been taken up in PR #4

**Pull request merged**